// es-abstract/2024/Number/index.js
'use strict';

import add from './add';
import bitwiseAND from './bitwiseAND';
import bitwiseNOT from './bitwiseNOT';
import bitwiseOR from './bitwiseOR';
import bitwiseXOR from './bitwiseXOR';
import divide from './divide';
import equal from './equal';
import exponentiate from './exponentiate';
import leftShift from './leftShift';
import lessThan from './lessThan';
import multiply from './multiply';
import remainder from './remainder';
import sameValue from './sameValue';
import sameValueZero from './sameValueZero';
import signedRightShift from './signedRightShift';
import subtract from './subtract';
import toString from './toString';
import unaryMinus from './unaryMinus';
import unsignedRightShift from './unsignedRightShift';

module.exports = {
	add: add,
	bitwiseAND: bitwiseAND,
	bitwiseNOT: bitwiseNOT,
	bitwiseOR: bitwiseOR,
	bitwiseXOR: bitwiseXOR,
	divide: divide,
	equal: equal,
	exponentiate: exponentiate,
	leftShift: leftShift,
	lessThan: lessThan,
	multiply: multiply,
	remainder: remainder,
	sameValue: sameValue,
	sameValueZero: sameValueZero,
	signedRightShift: signedRightShift,
	subtract: subtract,
	toString: toString,
	unaryMinus: unaryMinus,
	unsignedRightShift: unsignedRightShift
};
