// app-builder-lib/out/targets/nsis/nsisLang.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addCustomMessageFileInclude = exports.createAddLangsMacro = exports.LangConfigurator = void 0;
import builder_util_1 from 'builder-util';
import langs_1 from '../../util/langs';
import debug_1 from 'debug';
import fs_extra_1 from 'fs-extra';
import js_yaml_1 from 'js-yaml';
import path from 'path';
import nsisUtil_1 from './nsisUtil';
const debug = (0, debug_1.default)("electron-builder:nsis");
class LangConfigurator {
    constructor(options) {
        const rawList = options.installerLanguages;
        if (options.unicode === false || rawList === null || (Array.isArray(rawList) && rawList.length === 0)) {
            this.isMultiLang = false;
        }
        else {
            this.isMultiLang = options.multiLanguageInstaller !== false;
        }
        if (this.isMultiLang) {
            this.langs = rawList == null ? langs_1.bundledLanguages.slice() : (0, builder_util_1.asArray)(rawList).map(it => (0, langs_1.toLangWithRegion)(it.replace("-", "_")));
        }
        else {
            this.langs = ["en_US"];
        }
    }
}
exports.LangConfigurator = LangConfigurator;
function createAddLangsMacro(scriptGenerator, langConfigurator) {
    const result = [];
    for (const langWithRegion of langConfigurator.langs) {
        let name;
        if (langWithRegion === "zh_CN") {
            name = "SimpChinese";
        }
        else if (langWithRegion === "zh_TW") {
            name = "TradChinese";
        }
        else if (langWithRegion === "nb_NO") {
            name = "Norwegian";
        }
        else if (langWithRegion === "pt_BR") {
            name = "PortugueseBR";
        }
        else {
            const lang = langWithRegion.substring(0, langWithRegion.indexOf("_"));
            name = langs_1.langIdToName[lang];
            if (name == null) {
                throw new Error(`Language name is unknown for ${lang}`);
            }
            if (name === "Spanish") {
                name = "SpanishInternational";
            }
        }
        result.push(`!insertmacro MUI_LANGUAGE "${name}"`);
    }
    scriptGenerator.macro("addLangs", result);
}
exports.createAddLangsMacro = createAddLangsMacro;
async function writeCustomLangFile(data, packager) {
    const file = await packager.getTempFile("messages.nsh");
    await (0, fs_extra_1.outputFile)(file, data);
    return file;
}
async function addCustomMessageFileInclude(input, packager, scriptGenerator, langConfigurator) {
    const data = (0, js_yaml_1.load)(await (0, fs_extra_1.readFile)(path.join(nsisUtil_1.nsisTemplatesDir, input), "utf-8"));
    const instructions = computeCustomMessageTranslations(data, langConfigurator).join("\n");
    debug(instructions);
    scriptGenerator.include(await writeCustomLangFile(instructions, packager));
}
exports.addCustomMessageFileInclude = addCustomMessageFileInclude;
function computeCustomMessageTranslations(messages, langConfigurator) {
    const result = [];
    const includedLangs = new Set(langConfigurator.langs);
    for (const messageId of Object.keys(messages)) {
        const langToTranslations = messages[messageId];
        const unspecifiedLangs = new Set(langConfigurator.langs);
        for (const lang of Object.keys(langToTranslations)) {
            const langWithRegion = (0, langs_1.toLangWithRegion)(lang);
            if (!includedLangs.has(langWithRegion)) {
                continue;
            }
            const value = langToTranslations[lang];
            if (value == null) {
                throw new Error(`${messageId} not specified for ${lang}`);
            }
            result.push(`LangString ${messageId} ${langs_1.lcid[langWithRegion]} "${value.replace(/\n/g, "$\\r$\\n")}"`);
            unspecifiedLangs.delete(langWithRegion);
        }
        if (langConfigurator.isMultiLang) {
            const defaultTranslation = langToTranslations.en.replace(/\n/g, "$\\r$\\n");
            for (const langWithRegion of unspecifiedLangs) {
                result.push(`LangString ${messageId} ${langs_1.lcid[langWithRegion]} "${defaultTranslation}"`);
            }
        }
    }
    return result;
}
