// semver/index.js
'use strict'

// just pre-load all the stuff that index.js lazily exports
import internalRe from './internal/re';
import constants from './internal/constants';
import SemVer from './classes/semver';
import identifiers from './internal/identifiers';
import parse from './functions/parse';
import valid from './functions/valid';
import clean from './functions/clean';
import inc from './functions/inc';
import diff from './functions/diff';
import major from './functions/major';
import minor from './functions/minor';
import patch from './functions/patch';
import prerelease from './functions/prerelease';
import compare from './functions/compare';
import rcompare from './functions/rcompare';
import compareLoose from './functions/compare-loose';
import compareBuild from './functions/compare-build';
import sort from './functions/sort';
import rsort from './functions/rsort';
import gt from './functions/gt';
import lt from './functions/lt';
import eq from './functions/eq';
import neq from './functions/neq';
import gte from './functions/gte';
import lte from './functions/lte';
import cmp from './functions/cmp';
import coerce from './functions/coerce';
import Comparator from './classes/comparator';
import Range from './classes/range';
import satisfies from './functions/satisfies';
import toComparators from './ranges/to-comparators';
import maxSatisfying from './ranges/max-satisfying';
import minSatisfying from './ranges/min-satisfying';
import minVersion from './ranges/min-version';
import validRange from './ranges/valid';
import outside from './ranges/outside';
import gtr from './ranges/gtr';
import ltr from './ranges/ltr';
import intersects from './ranges/intersects';
import simplifyRange from './ranges/simplify';
import subset from './ranges/subset';
module.exports = {
  parse,
  valid,
  clean,
  inc,
  diff,
  major,
  minor,
  patch,
  prerelease,
  compare,
  rcompare,
  compareLoose,
  compareBuild,
  sort,
  rsort,
  gt,
  lt,
  eq,
  neq,
  gte,
  lte,
  cmp,
  coerce,
  Comparator,
  Range,
  satisfies,
  toComparators,
  maxSatisfying,
  minSatisfying,
  minVersion,
  validRange,
  outside,
  gtr,
  ltr,
  intersects,
  simplifyRange,
  subset,
  SemVer,
  re: internalRe.re,
  src: internalRe.src,
  tokens: internalRe.t,
  SEMVER_SPEC_VERSION: constants.SEMVER_SPEC_VERSION,
  RELEASE_TYPES: constants.RELEASE_TYPES,
  compareIdentifiers: identifiers.compareIdentifiers,
  rcompareIdentifiers: identifiers.rcompareIdentifiers,
}
