// electron-builder/out/cli/install-app-deps.js
#! /usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installAppDeps = exports.configureInstallAppDepsCommand = void 0;
import electronVersion_1 from 'app-builder-lib/out/electron/electronVersion';
import config_1 from 'app-builder-lib/out/util/config';
import yarn_1 from 'app-builder-lib/out/util/yarn';
import version_1 from 'app-builder-lib/out/version';
import packageDependencies_1 from 'app-builder-lib/out/util/packageDependencies';
import builder_util_1 from 'builder-util';
import promise_1 from 'builder-util/out/promise';
import fs_extra_1 from 'fs-extra';
import lazy_val_1 from 'lazy-val';
import path from 'path';
import read_config_file_1 from 'read-config-file';
import yargs from 'yargs';
/** @internal */
function configureInstallAppDepsCommand(yargs) {
    // https://github.com/yargs/yargs/issues/760
    // demandOption is required to be set
    return yargs
        .parserConfiguration({
        "camel-case-expansion": false,
    })
        .option("platform", {
        choices: ["linux", "darwin", "win32"],
        default: process.platform,
        description: "The target platform",
    })
        .option("arch", {
        choices: (0, builder_util_1.getArchCliNames)().concat("all"),
        default: process.arch === "arm" ? "armv7l" : process.arch,
        description: "The target arch",
    });
}
exports.configureInstallAppDepsCommand = configureInstallAppDepsCommand;
/** @internal */
async function installAppDeps(args) {
    try {
        builder_util_1.log.info({ version: version_1.PACKAGE_VERSION }, "electron-builder");
    }
    catch (e) {
        // error in dev mode without babel
        if (!(e instanceof ReferenceError)) {
            throw e;
        }
    }
    const projectDir = process.cwd();
    const packageMetadata = new lazy_val_1.Lazy(() => (0, read_config_file_1.orNullIfFileNotExist)((0, fs_extra_1.readJson)(path.join(projectDir, "package.json"))));
    const config = await (0, config_1.getConfig)(projectDir, null, null, packageMetadata);
    const [appDir, version] = await Promise.all([
        (0, config_1.computeDefaultAppDirectory)(projectDir, (0, builder_util_1.use)(config.directories, it => it.app)),
        (0, electronVersion_1.getElectronVersion)(projectDir, config),
    ]);
    // if two package.json — force full install (user wants to install/update app deps in addition to dev)
    await (0, yarn_1.installOrRebuild)(config, appDir, {
        frameworkInfo: { version, useCustomDist: true },
        platform: args.platform,
        arch: args.arch,
        productionDeps: (0, packageDependencies_1.createLazyProductionDeps)(appDir, null),
    }, appDir !== projectDir);
}
exports.installAppDeps = installAppDeps;
function main() {
    return installAppDeps(configureInstallAppDepsCommand(yargs).argv);
}
if (require.main === module) {
    builder_util_1.log.warn("please use as subcommand: electron-builder install-app-deps");
    main().catch(promise_1.printErrorAndExit);
}
